### PR TITLE
Split the `app` package

### PR DESCRIPTION
**Pull request by Idloj**
Note: `depend.graffle` and `depend.pdf` should be updated to include the recent changes (the new `ide` package, and this PR)

### Before

```
app
├── AboutWindow.scala
├── AgentMonitorEditor.scala
├── AgentMonitorManager.scala
├── AgentMonitor.scala
├── AgentMonitorView.java
├── AgentMonitorViewPanel.scala
├── AgentMonitorWindow.scala
├── App.scala
├── CodeTab.scala
├── CodeToHtmlInterface.scala
├── CodeToHtml.scala
├── CommandCenter.scala
├── CommandLine.scala
├── DirtyMonitor.scala
├── EditMenu.scala
├── EditorAreaWrapper.scala
├── EditorFactory.scala
├── ExtensionAssistant.scala
├── FileMenu.scala
├── FindDialog.java
├── GraphicsPreview.scala
├── HelpMenu.scala
├── HistoryPrompt.scala
├── IncludesMenu.scala
├── InfoFormatter.scala
├── InfoTab.scala
├── InterfacePanel.scala
├── InterfaceTab.scala
├── InterfaceToolBar.scala
├── LinePrompt.scala
├── MainCodeTab.scala
├── ModelSaver.scala
├── ModelsLibraryDialog.java
├── ModelsLibraryIndexReader.scala
├── NetLogoWebSaver.scala
├── Plugins.scala
├── Preference.scala
├── PreferencesDialog.scala
├── previewcommands
│   ├── EditorPanel.scala
│   ├── GUIState.scala
│   ├── PreviewCommandsDialog.scala
│   ├── PreviewCommandsEditor.scala
│   └── PreviewPanel.scala
├── ProceduresMenu.scala
├── ProceduresMenuTarget.scala
├── RecentFilesMenu.scala
├── SmartIndenter.scala
├── Splash.scala
├── Tabs.scala
├── TaskWarning.scala
├── TemporaryCodeTab.scala
├── ToolsMenu.scala
├── WidgetCreator.scala
├── WidgetPanel.scala
├── WidgetWrapper.java
├── WindowBar.java
└── ZoomMenu.scala
```

### After

```
app
├── AboutWindow.scala
├── App.scala
├── codetab
│   ├── CodeTab.scala
│   ├── EditorAreaWrapper.scala
│   ├── IncludesMenu.scala
│   ├── MainCodeTab.scala
│   ├── ProceduresMenu.scala
│   ├── ProceduresMenuTarget.scala
│   ├── SmartIndenter.scala
│   └── TemporaryCodeTab.scala
├── CodeToHtml.scala
├── common
│   ├── CommandLine.scala
│   ├── EditorFactory.scala
│   ├── FindDialog.java
│   ├── HistoryPrompt.scala
│   ├── LinePrompt.scala
│   └── TabsInterface.scala
├── DirtyMonitor.scala
├── EditMenu.scala
├── ExtensionAssistant.scala
├── FileMenu.scala
├── HelpMenu.scala
├── infotab
│   ├── InfoFormatter.scala
│   └── InfoTab.scala
├── interfacetab
│   ├── CommandCenter.scala
│   ├── InterfacePanel.scala
│   ├── InterfaceTab.scala
│   ├── InterfaceToolBar.scala
│   ├── WidgetCreator.scala
│   ├── WidgetPanel.scala
│   ├── WidgetWrapperEdge.java
│   └── WidgetWrapper.java
├── ModelSaver.scala
├── Plugins.scala
├── RecentFilesMenu.scala
├── Splash.scala
├── Tabs.scala
├── TaskWarning.scala
├── tools
│   ├── AgentMonitorEditor.scala
│   ├── AgentMonitorManager.scala
│   ├── AgentMonitor.scala
│   ├── AgentMonitorView.java
│   ├── AgentMonitorViewPanel.scala
│   ├── AgentMonitorWindow.scala
│   ├── EditorPanel.scala
│   ├── GraphicsPreview.scala
│   ├── GUIState.scala
│   ├── ModelsLibraryDialog.java
│   ├── ModelsLibraryIndexReader.scala
│   ├── NetLogoWebSaver.scala
│   ├── Preference.scala
│   ├── PreferencesDialog.scala
│   ├── PreviewCommandsDialog.scala
│   ├── PreviewCommandsEditor.scala
│   └── PreviewPanel.scala
├── ToolsMenu.scala
└── ZoomMenu.scala
```

### CODE REVIEW

File: project/Depend.scala [hunk 1]
```diff
@@ -41,7 +41,12 @@ object Depend {
       "" -> Nil,
       "agent" -> List("api"),
       "api" -> List("core", "core/model", "util"),
-      "app" -> List("ide", "window"),
+      "app" -> List("app/codetab", "app/common", "app/infotab", "app/interfacetab", "app/tools"),
+      "app/codetab" -> List("app/common"),
+      "app/infotab" -> List("app/common"),
+      "app/interfacetab" -> List("app/common", "app/tools"),
```
Comment (Idloj) [inline project/Depend.scala]: I realize the dependence on "app/tools" might not be desired. If not, let me know and I'll cut this dependency out (it's very easy to)

File: netlogo-gui/src/main/app/codetab/IncludesMenu.scala [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  override def populate(menu: JPopupMenu) {
+    target.getIncludesTable match {
+      case Some(includesTable) =>
+        this.includesTable = includesTable
+        val includes = new JArrayList[String]
+        includesTable.keys.foreach(includes.add)
+        Collections.sort(includes, String.CASE_INSENSITIVE_ORDER)
+        for(include <- includes.asScala if include.endsWith(".nls") && include.size > 4)
```
Comment (mrerrormessage) [inline netlogo-gui/src/main/app/codetab/IncludesMenu.scala]: We shouldn't be using a `JArrayList` here. Let's use `includesTable.keys.toSeq.sortBy(_.toUpperCase)`.

File: project/Depend.scala (hunk 1 above)
Comment (mrerrormessage) [inline project/Depend.scala]: I don't see a problem with it.

**Comment by mrerrormessage** [comment]
Let's remove `app/CodeToHtmlInterface.scala`. Other than that, this is good to go. Thanks for working on this!

**Comment by mrerrormessage** [comment]
Also, minor quibble, can you modify your commit message and leave a line of whitespace between the first line ("Split the `app` package") and the second line ("Files (not all of them)..."). It makes it show up more nicely in git.

**Comment by Idloj** [comment]
Side question: should I do it for all commit messages (that contain more than one line) from now on?

File: extensions/vid [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit c440a5ff849118d42143dc0fb649127b92e3dbc0
+Subproject commit 39005161a92705493c906698d097e6df64d6846a
```
Comment (Idloj) [inline extensions/vid]: How do I get rid of this change?

File: netlogo-gui/src/main/app/codetab/IncludesMenu.scala (hunk 2 above)
Comment (Idloj) [inline netlogo-gui/src/main/app/codetab/IncludesMenu.scala]: I don't mind opening a PR that changes that (and possibly other things in IncludesMenu)

**Comment by mrerrormessage** [comment]
>Side question: should I do it for all commit messages (that contain multiple lines) from now on?

That would be great!

> How do I get rid of this change?

`git checkout hexy -- extensions/vid` will check out the same revision of the submodule as in hexy. You can then `git add extensions/vid` and `git commit --amend` to get that change into your previous commit.

**Comment by Idloj** [comment]
Another question:
Should I move "Models Library Dialog" stuff to `app.tools`? What about NetLogoWebSaver?

**Comment by mrerrormessage** [comment]
If you can move them to `app.tools` without breaking any other classes, I say go for it.

**Comment by Idloj** [comment]
Moved them to `app.tools`.

I can't get rid of the vid extension change, it doesn't seem to work.

**Comment by mrerrormessage** [comment]
The `extensions/vid` change doesn't show up in the PR, so it's fine if it changed on your branch because it won't change when this gets merged in.

**Comment by Idloj** [comment]
yeah, I figured out how to get rid of the change

**Comment by Idloj** [comment]
I think this is ready for merge

**Comment by mrerrormessage** [comment]
Looks great! Thanks for doing this!

**Pull request merged**